dChatServer/ChatPacketHandler.cpp

```diff
@@ -398,72 +386,83 @@ void ChatPacketHandler::HandlePrivateChatMessage(Packet* packet) {
 
 	eChatChannel channel;
 	uint32_t size;
-	LUWString receiverName;
+	LUWString LUReceiverName;
 
-	inStream.IgnoreBytes(4);
+	inStream.IgnoreBytes(4); // padding/garbage
 	inStream.Read(channel);
+	if (channel != eChatChannel::PRIVATE_CHAT) LOG("WARNING: Received Private chat with the wrong channel!");
+
 	inStream.Read(size);
-	inStream.IgnoreBytes(77);
-	inStream.Read(receiverName);
-	inStream.IgnoreBytes(2);
+	inStream.IgnoreBytes(77); // padding/garbage
+
+	inStream.Read(LUReceiverName);
+	auto receiverName = LUReceiverName.GetAsString();
+	inStream.IgnoreBytes(2); // padding/garbage
 
 	LUWString message(size);
 	inStream.Read(message);
 	
-	LOG("Got a message from (%s) via [%s]: %s to %s", sender.playerName.c_str(), StringifiedEnum::ToString(channel).data(), message.GetAsString().c_str(), receiverName.GetAsString().c_str());
+	LOG("Got a message from (%s) via [%s]: %s to %s", sender.playerName.c_str(), StringifiedEnum::ToString(channel).data(), message.GetAsString().c_str(), receiverName.c_str());
 
-	const auto& receiver = Game::playerContainer.GetPlayerData(receiverName.GetAsString());
-	if (!receiver) return; // TODO: respond with not online or error
+	const auto& receiver = Game::playerContainer.GetPlayerData(receiverName);
+	if (!receiver) {
+		PlayerData otherPlayer;
+		otherPlayer.playerName = receiverName;
+		auto responseType = Database::Get()->GetCharacterInfo(receiverName)
+			? eChatMessageResponseCode::NOTONLINE
+			: eChatMessageResponseCode::GENERALERROR;
 
-	// TODO: check if they are friends, only friends should be able to whisper/private chat
+		SendPrivateChatMessage(sender, otherPlayer, sender, message, responseType);
+		return;
+	}
 
-	//To the sender:
-	{
-		CBITSTREAM;
-		BitStreamUtils::WriteHeader(bitStream, eConnectionType::CHAT_INTERNAL, eChatInternalMessageType::ROUTE_TO_PLAYER);
-		bitStream.Write(sender.playerID);
+	if (sender.isFTP) {
+		SendPrivateChatMessage(sender, receiver, sender, message, eChatMessageResponseCode::SENDERFREETRIAL);
+		return;
+	}
 
-		BitStreamUtils::WriteHeader(bitStream, eConnectionType::CHAT, eChatMessageType::PRIVATE_CHAT_MESSAGE);
-		bitStream.Write(sender.playerID);
-		bitStream.Write(eChatChannel::PRIVATE_CHAT);
-		bitStream.Write<uint32_t>(0); // not used
-		bitStream.Write(LUWString(sender.playerName));
-		bitStream.Write(sender.playerID);
-		bitStream.Write<uint16_t>(0); // sourceID
-		bitStream.Write(sender.GMLevel);
-		bitStream.Write(LUWString(receiver.playerName));
-		bitStream.Write(receiver.GMLevel);
-		bitStream.Write(eChatMessageResponseCode::Sent);
-		bitStream.Write(message);
+	if (receiver.isFTP) {
+		SendPrivateChatMessage(sender, receiver, sender, message, eChatMessageResponseCode::RECEIVERFREETRIAL);
+		return;
+	}
```

Comment:
Same thing here.  I dont like the prospect of actually starting to add code to support FTP features and locks.